### PR TITLE
Multi Token Pool

### PR DESCRIPTION
**Pull request by jacque006**
Resolves https://github.com/thehubbleproject/hubble-contracts/issues/626

### TODO
- [x] Unit test `MultiTokenPool`
- [x] Add multiple tokens to client integration tests
~Add script to generate and deposit (create state) for fee receivers~ https://github.com/thehubbleproject/hubble-contracts/issues/476
- [x] Double check `number` -> `BigNumber` refactor points.

### CODE REVIEW

File: scripts/deploy.ts [hunk 1]
```diff
@@ -80,10 +80,11 @@ async function main() {
             `Registering ${numPubkeys} pubkeys. Custom mnemonic: ${!!pubkeyMnemonic}`
         );
         const group = Group.new({
-            n: pubkeyMnemonic,
+            n: parseInt(numPubkeys),
```
Comment (jacque006) [inline scripts/deploy.ts:83]: Pretty sure this derp was me, should work properly now.

File: ts/client/features/deposit.ts [hunk 2]
```diff
@@ -116,7 +116,7 @@ export async function handleNewBatch(
     const receipt = await event.getTransactionReceipt();
     const logs = receipt.logs.map(log => rollup.interface.parseLog(log));
     const depositsFinalisedLog = logs.filter(
-        log => log.signature == "DepositsFinalised(uint256,bytes32,uint256)"
+        log => log.signature === "DepositsFinalised(uint256,bytes32,uint256)"
```
Comment (jacque006) [inline ts/client/features/deposit.ts:119]: `===` > `==` https://stackoverflow.com/a/3586781

File: ts/client/storageEngine/interfaces.ts [hunk 3]
```diff
@@ -10,6 +10,8 @@ export interface Entry<Item> {
     item: Item;
 }
 
+// TODO Update this interface and implementations to
+// use BigNumber for itemID.
```
Comment (jacque006) [inline ts/client/storageEngine/interfaces.ts:14]: Avoided this refactor as this PR is already large and didn't want to bork @kautukkundan 's work :) `BigNumberish` should be fine as well as long as the storage engine and tree both store in a large value format under the covers.

File: ts/mcl.ts [hunk 4]
```diff
@@ -176,7 +176,7 @@ export function parseG2(solG2: solG2): mclG2 {
 }
 
 function dump(sol: solG1 | solG2): string {
-    return `0x${sol.map(n => n.toString().slice(2)).join()}`;
+    return `0x${sol.map(n => n.toString().slice(2)).join("")}`;
```
Comment (jacque006) [inline ts/mcl.ts:179]: This is also my derp. `join` by default adds a comma separator which the previous implementation did not.

**Comment by jacque006** [comment]
@kautukkundan @ChihChengLiang your :eyes: would be especially appreciated on `number` -> `BigNumber` refactor points. I avoided updating code in `commitments.ts` and `StorageEngine`, mainly focused on values in `ts/features/transfers`.

File: contracts/test/CustomToken.sol [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * @dev Contract for use testing multi-token scenarios.
+ */
+contract CustomToken is ERC20, Ownable {
+    constructor(string memory name, string memory symbol)
+        public
+        ERC20(name, symbol)
+    {
+        _mint(msg.sender, 1000000000000000000000000000);
```
Comment (kautukkundan) [inline contracts/test/CustomToken.sol:15]: can use exponential notation `1e27` for simplicity

File: contracts/test/ExampleToken.sol [hunk 6]
```diff
@@ -6,7 +6,8 @@ import "@openzeppelin/contracts/access/Ownable.sol";
 
 contract ExampleToken is ERC20, Ownable {
     /**
-     * @dev assign totalSupply to account creating this contract */
+     * @dev assign totalSupply to account creating this contract
+     */
     constructor() public ERC20("Example", "EMP") {
         _mint(msg.sender, 1000000000000000000000000000);
```
Comment (kautukkundan) [inline contracts/test/ExampleToken.sol:12]: same exponential notation can be used here

File: test/client/integration.test.ts [hunk 7]
```diff
@@ -60,49 +61,69 @@ describe("Client Integration", function() {
 
         const contracts = await deployAll(signer, parameters);
 
+        // Setup and register custom tokens
+        const customTokens = await Promise.all([
+            new CustomToken__factory(signer).deploy("Hubble", "HUB"),
+            new CustomToken__factory(signer).deploy("Telescope", "TLSC")
```
Comment (kautukkundan) [inline test/client/integration.test.ts:67]: Telescope token ❤️

File: test/client/transaction.test.ts [hunk 8]
```diff
@@ -13,6 +14,22 @@ import {
 
 chai.use(chaiAsPromised);
 
+const txFactory = (
```
Comment (kautukkundan) [inline test/client/transaction.test.ts:17]: 👍🏼

File: ts/client/storageEngine/interfaces.ts (hunk 3 above)
Comment (kautukkundan) [reply ts/client/storageEngine/interfaces.ts:14]: oh makes sense, can discuss this

File: ts/client/config.ts [hunk 9]
```diff
@@ -1,12 +1,13 @@
+import { BigNumberish } from "ethers";
 import { readFile } from "fs";
 import { resolve } from "path";
 import { promisify } from "util";
 
 const readFileAsync = promisify(readFile);
 
 export type FeeReceivers = Array<{
```
Comment (ChihChengLiang) [inline ts/client/config.ts:8]: What do you think we make FeeReceivers a class, so we can build the get random token getter or iterators from the object method?

File: test/fast/txPool.test.ts [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+    tokenID: BigNumber;
+    feeReceiverID: BigNumber;
+    sumFees: BigNumber;
+}): { tokenID: number; feeReceiverID: number; sumFees: number } => ({
+    tokenID: tokenID.toNumber(),
+    feeReceiverID: feeReceiverID.toNumber(),
+    sumFees: sumFees.toNumber()
+});
+
+const assertTXEq = (a: TransferOffchainTx, b: TransferOffchainTx) => {
```
Comment (ChihChengLiang) [inline test/fast/txPool.test.ts:65]: We can also compare the tx hashes
`assert.equal(a.hash(), b.hash())`

File: ts/client/txPool.ts [hunk 11]
```diff
@@ -1,40 +1,159 @@
...
+        ];
+        return {
+            tokenID,
+            feeReceiverID,
+            sumFees: highValue
+        };
+    }
+
+    private getQueueValue(tokenIDStr: string): BigNumber {
+        return sum(this.tokenIDStrToQueue[tokenIDStr].map(tx => tx.fee));
```
Comment (ChihChengLiang) [inline ts/client/txPool.ts:157]: Note that we have a queue size of 1024.
We have a commitment size of 32.
We pack 1 commitment at a time. 
So if we have a token A queue of 32 txs each worth $10 and a token B queue of 1024 tx each worth $1. The `getQueueValue ` considers token B queue is more valuable but token A queue is more profitable in a single pack. 

Not a big problem to fix now though.

**Review by ChihChengLiang — COMMENTED** [review]
Looks good to me

File: ts/client/config.ts (hunk 9 above)
Comment (jacque006) [reply ts/client/config.ts:8]: We can do that, I see the random part as more test related but no harm in having that available in prod code.

File: ts/client/txPool.ts (hunk 11 above)
Comment (jacque006) [reply ts/client/txPool.ts:157]: Ah good catch, that is something I can easily fix.

Comment (jacque006) [reply ts/client/txPool.ts:157]: Move to new issue

File: ts/client/config.ts (hunk 9 above)
Comment (jacque006) [reply ts/client/config.ts:8]: https://github.com/thehubbleproject/hubble-contracts/issues/628

File: ts/client/txPool.ts (hunk 11 above)
Comment (jacque006) [reply ts/client/txPool.ts:157]: https://github.com/thehubbleproject/hubble-contracts/issues/628

**Pull request merged**